include/onnxruntime/core/common/string_helper.h

```diff
@@ -6,7 +6,7 @@
 #include "core/session/onnxruntime_c_api.h"
 
 namespace onnxruntime {
-#if defined(USE_TENSORRT) || defined(USE_CUDA)
+#if defined(USE_TENSORRT) || defined(USE_CUDA) || defined(ENABLE_TRAINING_ON_DEVICE)
 static char* StrDup(const std::string& str, _Inout_ OrtAllocator* allocator) {
```

Comment:
>static cha [](http://example.com/codeflow?start=0&length=10)

There are few issues here:
1) We do not want public APIs to be include into common. Then the whole codebase potentially includes public headers. API supposed to be a front interface.
2) This function should be a util, perhaps in the core/common/session and you can include it from there
3) Function bodies in the headers needs to be inline otherwise every file that includes it has its own definition, and the original purpose was to avoid code duplication.
4) I suggest to put this into a CC file.